support/langkit_support-generic_api-analysis.ads

```diff
@@ -0,0 +1,270 @@
+      File_Reader   : File_Reader_Reference := No_File_Reader_Reference;
+      With_Trivia   : Boolean := True;
+      Tab_Stop      : Natural := 0)
+      return Analysis_Context;
+   --  Create a new analysis context.
+   --
+   --  ``Charset`` will be used as a default charset to decode input sources
+   --  in analysis units. Please see ``GNATCOLL.Iconv`` for several supported
+   --  charsets. Be careful: passing an unsupported charset is not guaranteed
+   --  to raise an error here. If no charset is provided, use the language
+   --  default.
+   --
+   --  When ``With_Trivia`` is true, the parsed analysis units will contain
+   --  trivias.
+   --
+   --  If provided, ``File_Reader`` will be used to fetch the contents of
+   --  source files instead of the default, which is to just read it from the
+   --  filesystem and decode it using the regular charset rules. Note that if
+   --  provided, all parsing APIs that provide a buffer are forbidden, and any
+   --  use of the rewriting API with the returned context is rejected.
+   --
+   --  ``Tab_Stop`` is a positive number to describe the effect of tabulation
+   --  characters on the column number in source files.
+
+   function Language_For (Context : Analysis_Context'Class) return Language_Id;
+   --  Return the unique identifier for the Langkit-generated library that
+   --  implements ``Context``.
+
+   function Has_Unit
+     (Context       : Analysis_Context'Class;
+      Unit_Filename : String) return Boolean;
+   --  Return whether ``Context`` contains a unit correponding to
+   --  ``Unit_Filename``.
+
+   function Get_From_File
+     (Context  : Analysis_Context'Class;
+      Filename : String;
+      Charset  : String := "";
+      Reparse  : Boolean := False;
+      Rule     : Any_Grammar_Rule := No_Grammar_Rule) return Analysis_Unit;
+   --  Create a new analysis unit for ``Filename`` or return the existing one
+   --  if any. If ``Reparse`` is true and the analysis unit already exists,
+   --  reparse it from ``Filename``.
+   --
+   --  ``Rule`` controls which grammar rule is used to parse the unit.
+   --
+   --  Use ``Charset`` in order to decode the source. If ``Charset`` is empty
+   --  then use the context's default charset.
+   --
+   --  If any failure occurs, such as file opening, decoding, lexing or parsing
+   --  failure, return an analysis unit anyway: errors are described as
+   --  diagnostics of the returned analysis unit.
+   --
+   --  It is invalid to pass ``True`` to ``Reparse`` if a rewriting context is
+   --  active.
+
+   --  TODO??? Bind all other analysis context primitives
+
+   ------------------------------
+   -- Analysis unit operations --
+   ------------------------------
+
+   overriding function Get_Line
+     (Self : Analysis_Unit; Line_Number : Positive) return Text_Type;
+   --  Return the line of text at line number ``Line_Number``
+
+   function Root (Unit : Analysis_Unit'Class) return Analysis_Node;
```

Comment:
I think having a link back to the Unit would be good here, I could see that as being generally useful in the generic API.

Reply:
Sure, I’ll also add the unit -> context backlink